cli/dcoscli/node/main.py

```diff
@@ -65,12 +72,303 @@ def _cmds():
             function=_ssh),
 
         cmds.Command(
+            hierarchy=['node', 'snapshot', 'create'],
+            arg_keys=['<nodes>'],
+            function=_snapshot_create),
+
+        cmds.Command(
+            hierarchy=['node', 'snapshot', 'delete'],
+            arg_keys=['<snapshot>'],
+            function=_snapshot_delete),
+
+        cmds.Command(
+            hierarchy=['node', 'snapshot', 'download'],
+            arg_keys=['<snapshot>', '--location'],
+            function=_snapshot_download),
+
+        cmds.Command(
+            hierarchy=['node', 'snapshot'],
+            arg_keys=['--list', '--status', '--cancel', '--json'],
+            function=_snapshot_manage),
+
+        cmds.Command(
             hierarchy=['node'],
             arg_keys=['--json'],
-            function=_list),
+            function=_list)
     ]
 
 
+def snapshot_error(fn):
+    @functools.wraps(fn)
+    def check_for_snapshot_error(*args, **kwargs):
+        response = fn(*args, **kwargs)
+        if response.status_code != 200:
+            err_msg = ('Error making {} request\nURL: '
+                       '{}, status_code: {}.'.format(args[1], args[0],
+                                                     response.status_code))
+            if not kwargs.get('stream'):
+                err_status = _read_http_response_body(response).get('status')
+                if err_status:
+                    err_msg = err_status
+            raise DCOSException(err_msg)
+        return response
+    return check_for_snapshot_error
+
+
+def _check_3dt_version():
+    if not http.check_cluster_capability('SUPPORT_CLUSTER_REPORT'):
+        raise DCOSException(
+            'DC/OS backend does not support snapshot capabilities in this '
+            'version. Must be DC/OS >= 1.8')
+
+
+def sizeof_fmt(num, suffix='B'):
+    for unit in ['', 'Ki', 'Mi', 'Gi', 'Ti', 'Pi', 'Ei', 'Zi']:
+        if abs(num) < 1024.0:
+            return "%3.1f%s%s" % (num, unit, suffix)
+        num /= 1024.0
+    return "%.1f%s%s" % (num, 'Yi', suffix)
+
+
+def _get_snapshots_json():
+    '''
+    Get a json with a list of snapshots
+
+    :return: available snapshots on a cluster.
+    :rtype: dict
+    '''
+    return _do_snapshot_request(SNAPSHOT_BASE_URL + '/list/all',
+                                'GET')
+
+
+def _get_snapshots_list():
+    '''
+    Get a list of tuples (snapshot_file_name, file_size), ..
+
+    :return: list of snapshots
+    :rtype: list of tuples
+    '''
+    available_snapshots = []
+    for _, snapshot_files in _get_snapshots_json().items():
+        if snapshot_files is None:
```

Comment:
this is a what backend sends as response if a master node does not have any snapshot available